### PR TITLE
Datadog Integration Acceptance Tests / Bug fixes

### PR DESCRIPTION
**Pull request by natemollica-nm**
### Changes proposed in this PR ###  
- Introduce datadog integration acceptance testing with Datadog API 
  - Consul Integration
  - DogstatsD (Unix Domain Sockets)
  - DogstatsD (UDP)
  - Openmetrics Prometheus
  - OTLP Collection Integration (HTTP)
  - OTLP Collection Integration (gRPC)
- Fix some bugs surrounding the Openmetrics and Consul Integration Checks
  - Consul Checks also require `prometheus.io` annotations to be removed while enabled to avoid conflicts w/ dd-agent
  -  Fix Openmetrics and Consul Integration check URLs to use `consul.fullname-server` helper template for accurate consul API access
  - Fix OTLP collection for `telemetry-collector-deployment.yaml` to include `OTEL_EXPORTER_OTLP_ENDPOINT` env variable for proper Datadog tag ingestion

### How I've tested this PR ###

- Ran: `make bats-test` for templating updates
- Ran acceptance tests for new Datadog framework and test scenarios following [CONTRIBUTING.md guide](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#using-acceptance-test-framework-to-debug).

### How I expect reviewers to test this PR ###
- Review acceptance tests and see if they're worth even having in the repository. They add a few more go dependencies for Datadog client API access
  - The tests access the datadog agent's local API with datadog dogstatsd stats enabled to parse metrics output from the collection event.
  - This testing leverages the Datadog [fake-intake](https://github.com/DataDog/datadog-agent/tree/main/test/fakeintake) image to simulate traffic ingestion and remove the requirement to have to use an API and APP key.

### Checklist ###
- [X] Tests added
- [X] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: acceptance/tests/datadog/datadog_test.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+	var root Root
+	err = json.Unmarshal([]byte(metricsOutput), &root)
+	require.NoError(t, err)
+	for _, check := range root.RunnerStats.Checks.Openmetrics {
+		if strings.Contains(check.CheckID, "consul") {
+			require.Equal(t, ``, check.LastError)
+			break
+		}
+		continue
+	}
```
Comment (DanStough) [inline acceptance/tests/datadog/datadog_test.go:307]: Not strictly required, but I wonder if we can refactor most of these test cases into function that takes in dd Helm Values, a fixture name, and a validation callback. Those seem like the only differences at first glance.

**Review by DanStough — COMMENTED** [review]
Great work @natemollica-nm! Mostly minor comments. I think the only thing holding up my approval was skipping the test if the flag isn't enabled and the changes to the `DeployKustomize` function.

File: .changelog/3685.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+```release-note:improvement
+helm: acceptance tests added for datadog integration features of consul-k8s
+```
```
[the remark this answers is not in the document]
Comment (natemollica-nm) [reply .changelog/3685.txt]: Good point! Removed 👍🏼

File: acceptance/framework/consul/helm_cluster.go [hunk 3]
```diff
@@ -672,6 +680,14 @@ func createOrUpdateLicenseSecret(t *testing.T, client kubernetes.Interface, cfg
 	CreateK8sSecret(t, client, cfg, namespace, config.LicenseSecretName, config.LicenseSecretKey, cfg.EnterpriseLicense)
 }
 
+func CreateOrUpdateDatadogSecret(t *testing.T, client kubernetes.Interface, cfg *config.TestConfig, namespace string) {
```
[the remark this answers is not in the document]
Comment (natemollica-nm) [reply acceptance/framework/consul/helm_cluster.go]: https://github.com/hashicorp/consul-k8s/blob/91abb5d31b645933e40bda50c5c99748be018159/acceptance/framework/consul/helm_cluster.go#L683-L689

Updated 👍🏼

File: acceptance/framework/consul/helm_cluster.go [hunk 4]
```diff
@@ -783,3 +799,25 @@ func CreateK8sSecret(t *testing.T, client kubernetes.Interface, cfg *config.Test
 		_ = client.CoreV1().Secrets(namespace).Delete(context.Background(), secretName, metav1.DeleteOptions{})
 	})
 }
+
+func CreateMultiKeyK8sSecret(t *testing.T, client kubernetes.Interface, cfg *config.TestConfig, namespace, secretName string, secretMap map[string]string) {
```
[the remark this answers is not in the document]
Comment (natemollica-nm) [reply acceptance/framework/consul/helm_cluster.go]: https://github.com/hashicorp/consul-k8s/blob/91abb5d31b645933e40bda50c5c99748be018159/acceptance/framework/consul/helm_cluster.go#L803-L823
Updated 👍🏼

File: acceptance/framework/k8s/deploy.go [hunk 5]
```diff
@@ -69,8 +69,10 @@ func DeployKustomize(t *testing.T, options *k8s.KubectlOptions, noCleanupOnFailu
 		KubectlDeleteK(t, options, kustomizeDir)
 	})
 
-	// The timeout to allow for connect-init to wait for services to be registered by the endpoints controller.
-	RunKubectl(t, options, "wait", "--for=condition=available", "--timeout=5m", fmt.Sprintf("deploy/%s", deployment.Name))
+	if !strings.Contains(t.Name(), "Datadog") {
```
[the remark this answers is not in the document]
Comment (natemollica-nm) [reply acceptance/framework/k8s/deploy.go]: https://github.com/hashicorp/consul-k8s/blob/91abb5d31b645933e40bda50c5c99748be018159/acceptance/framework/k8s/deploy.go#L72-L73

Removed 👍🏼 -- I had this in here for some previous test runs using an actual dd-agent and Datadog API (vice the fake-intake). Retested after removal and tests ran fine. Good catch!

File: acceptance/framework/flags/flags.go [hunk 6]
```diff
@@ -198,14 +216,16 @@ func (t *TestFlags) TestConfigFromFlags() *config.TestConfig {
...
 	consulDataplaneVersion, _ := version.NewVersion(t.flagConsulDataplaneVersion)
-	//vaultserverVersion, _ := version.NewVersion(t.flagVaultServerVersion)
 	kubeEnvs := config.NewKubeTestConfigList(t.flagKubeconfigs, t.flagKubecontexts, t.flagKubeNamespaces)
 
 	c := &config.TestConfig{
-
 		EnableEnterprise:  t.flagEnableEnterprise,
 		EnterpriseLicense: t.flagEnterpriseLicense,
 
+		EnableDatadog: t.flagEnableDatadog,
```
[the remark this answers is not in the document]
Comment (natemollica-nm) [reply acceptance/framework/flags/flags.go:225]: Good catch! Updated 👍🏼 

https://github.com/hashicorp/consul-k8s/blob/91abb5d31b645933e40bda50c5c99748be018159/acceptance/framework/consul/helm_cluster.go#L97-L104

File: acceptance/tests/datadog/datadog_test.go [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+		"global.metrics.datadog.dogstatsd.dogstatsdTags[0]":    "source:consul",
+		"global.metrics.datadog.dogstatsd.dogstatsdTags[1]":    "consul_service:consul-server",
+		"global.metrics.datadog.dogstatsd.dogstatsdTags[2]":    acceptanceTestingTags,
+	}
+
+	datadogOperatorHelmValues := map[string]string{
+		"replicaCount":     "1",
+		"image.tag":        datadog.DefaultHelmChartVersion,
+		"image.repository": "gcr.io/datadoghq/operator",
+	}
```
[the remark this answers is not in the document]
Comment (natemollica-nm) [reply acceptance/tests/datadog/datadog_test.go]: Good point! 👍🏼 Updated with blank `map[string]string` in all tests.

https://github.com/hashicorp/consul-k8s/blob/91abb5d31b645933e40bda50c5c99748be018159/acceptance/tests/datadog/datadog_test.go#L63

File: acceptance/tests/datadog/datadog_test.go (hunk 1 above)
Comment (natemollica-nm) [reply acceptance/tests/datadog/datadog_test.go:284]: For sure! Yeah, this definitely could be refactored that way. I want to get this merged to get the naming convention issue fixed, so I left a TODO in there for now.

https://github.com/hashicorp/consul-k8s/blob/91abb5d31b645933e40bda50c5c99748be018159/acceptance/tests/datadog/datadog_test.go#L26

File: acceptance/framework/datadog/datadog.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	kubernetesClient kubernetes.Interface
+
+	noCleanupOnFailure bool
+	noCleanup          bool
+	debugDirectory     string
+	logger             terratestLogger.TestLogger
+}
+
+// releaseLabelSelector returns label selector that selects all pods
+// from a Vault installation.
```
[the remark this answers is not in the document]
Comment (DanStough) [reply acceptance/framework/datadog/datadog.go]: @natemollica-nm non-blocking but I think this is still a typo.

**Comment by natemollica-nm** [comment]
@DanStough Alright pushed the refactor in for removing the `-enable-datadog` flag, `-datadog-api-key`, and `-datadog-app-key`. Hard coded the secret creation in `acceptance/framework/datadog/datadog.go` and added in `-skip-datadog` flag to skip these tests if necessary.

File: acceptance/framework/flags/flags.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (natemollica-nm) [reply acceptance/framework/flags/flags.go]: Agreed on removing flag as it wasn't doing much in terms of controlling test runs.

File: acceptance/framework/config/config.go [hunk 9]
```diff
@@ -70,6 +70,9 @@ type TestConfig struct {
 	EnableEnterprise  bool
 	EnterpriseLicense string
 
+	SkipDataDogTests        bool
+	DatadogHelmChartVersion string
```
Comment (DanStough) [inline acceptance/framework/config/config.go:74]: 1. I think we can get rid of the `skipDataDogTests` flag altogether - the datadog tests won't run in CI unless we change the YAML we were looking at yesterday.
2. I think there is no way to set `DatadogHelmChartVersion`, so we can just remove it.

**Review by DanStough — APPROVED** [review]
Approving but please see the last comment - I trust you can resolve it before merging. Thanks again!

File: acceptance/framework/config/config.go (hunk 9 above)
Comment (DanStough) [reply acceptance/framework/config/config.go:74]: 😮‍💨 my bad, I didn't realize automerge was set.

Comment (natemollica-nm) [reply acceptance/framework/config/config.go:74]: Nah, that's my bad. Sorry 🫤

**Pull request merged**